### PR TITLE
Filter and de-dupe execution of outgoing pending operations

### PR DESCRIPTION
**Pull request by MehrdadMzfr**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [X] Are tests passing locally?
* [X] Are the files formatted correctly?
* [ ] Did you add unit tests? No, there is a task tracking it(to be done after the rebranding) https://msmobilecenter.visualstudio.com/Mobile-Center/_workitems/edit/60651 
* [X] Did you check UI tests on the sample app? They are not executed on CI.
* [X] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Filter and de-dupe execution of outgoing pending operations and ensure the operation is not ran multiple time.

### CODE REVIEW

File: AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m [hunk 1]
```diff
@@ -77,13 +77,16 @@ - (BOOL)upsertWithToken:(MSTokenResult *)token
   // This is the same as [[NSDate date] timeIntervalSince1970] - but saves us from allocating an NSDate.
   NSTimeInterval now = NSDate.timeIntervalSinceReferenceDate + NSTimeIntervalSince1970;
   NSString *tableName = [MSDBDocumentStore tableNameForPartition:token.partition];
+
+  // If operation is nil, pass NULL value, else user the operation name in this format '<OPERATION_NAME>' (note the single quotes).
+  NSString *normalizedOperationString = operation != nil ? [NSString stringWithFormat:@"'%@'", operation] : nil;
```
Comment (yochaude) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: Does passing `nil` here actually works? Or should you pass the string `"NULL"`? I am doing the later one somewhere else, but `nil` might just work.

Comment (MehrdadMzfr) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: tried it, it's working

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 2]
```diff
@@ -644,11 +644,24 @@ - (void)startWithChannelGroup:(id<MSChannelGroupProtocol>)channelGroup
     self.httpClient = [MSHttpClient new];
   }
 
-  // Listen to network events.
-  [MS_NOTIFICATION_CENTER addObserver:self selector:@selector(networkStateChanged:) name:kMSReachabilityChangedNotification object:nil];
+  // Listen to notificaitons.
```
Comment (clpolet) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]:
```suggestion
  // Listen to notifications.
```

Comment (clpolet) [reply AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: Although you shouldn't have to do it at start time. The service abstraction will actually call the `applyEnabledState` if the SDK is enabled while starting.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 3]
```diff
@@ -693,18 +710,18 @@ - (void)authTokenContext:(MSAuthTokenContext *)__unused authTokenContext didUpda
 
 #pragma mark - Reachability
 
-- (void)networkStateChanged:(NSNotificationCenter *)__unused notification {
+- (void)networkStateChangedDataStoreCallback:(NSNotificationCenter *)__unused notification {
```
Comment (clpolet) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]: The old name `networkStateChanged:` was OK to me, that's what we do in `MSHttpClient.m`.

File: AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m [hunk 4]
```diff
@@ -729,6 +746,19 @@ - (void)onNetworkGoesOnline {
...
+                                       __block NSString *operationId =
+                                           [MSDocumentUtils outgoingOperationIdWithPartition:kMSDataStoreUserDocumentsPartition
+                                                                                  documentId:operation.documentId];
+
+                                       // If the operation is already being processed, skip it.
+                                       if ([self.outgoingPendingOperations containsObject:operationId]) {
+                                         continue;
+                                       }
+
+                                       // Add current operaiton as pending.
```
Comment (clpolet) [inline AppCenterDataStorage/AppCenterDataStorage/MSDataStore.m]:
```suggestion
                                       // Add current operation as pending.
```

File: AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m (hunk 1 above)
Comment (clpolet) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: I would still use "NULL" for consistency.

Comment (clpolet) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: That's what we use in core too.

[the remark this answers is not in the document]
Comment (MehrdadMzfr) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/LocalStore/MSDBDocumentStore.m]: I verified this and  nil will become NULL when inserted in the table

**Pull request merged**